Simplify and robustify transport cargo killing and veterancy

clyfordv opened the pull request:
1. When killing cargo, the instigator unit is provided. The engine then counts kills for score and unitview purposes.
2. Similarly, the mass of all cargo is distributed to the killer(s) (via lua) with `VeterancyComponent.VeterancyDispersal`.
3. For transports with external storage, cargo is cached so impact effects can be delayed/applied properly when the transport hits the ground.
4. Lastly, the new function explicitly kills all units (previously relied on the engine to do it), which (hopefully) addresses a bug where units who were supposed to die escaped their fate.

And hopefully all in a much more straightforward way than before!

lL1l1 commented on lua/sim/units/components/TransportUnitComponent.lua:
```diff
@@ -59,30 +65,78 @@ BaseTransport = ClassSimple {
+
+            -- cache the cargo so we can impact it later (if needed)
+            -- exception for command units, which explode immediately
+            if not EntityCategoryContains(categories.COMMAND, unit) and cacheCargo then
+                unit.killedInTransport = true
+                table.insert(self.cargoCache, unit)
+            end
```
```suggestion
            if not EntityCategoryContains(categories.COMMAND, unit) then
                unit.killedInTransport = true
            end
            if cacheCargo then
                table.insert(self.cargoCache, unit)
            end
```
The two flags have independent purposes (triggering death effects from `Unit:Kill()` right now (on line 119) vs dealing with the units in a future time), so they should be treated separately.

clyfordv replied:
yep they gotta be separate

lL1l1 replied:
They're actually dependent on each other. killedInTransport means "don't trigger OnKilled effects right now because we will deal with them later" and cacheCargo means "these are the units we are dealing with later". So they should be under the same if statement.
> triggering death effects from Unit:Kill() right now
That should be a third option as I described before, specific for carriers to "kill units now, but without effects".

clyfordv replied:
I knew there was a reason I put them together ;D

But yes you're definitely right about the third option, I'll assemble it.

lL1l1 reviewed (changes requested):
I tested the changes with continentals of SACU or harbingers, a T1 transport with SACU/harbingers, a carrier full of strategic bombers, and a CZAR. The veterancy distribution works.
In the suggestions, I fixed two bugs I found:
- units don't detach from air transports nor spawn wrecks on transport death. They also play death effects when the transport dies.
- Units in carriers (transport internal storage) play death effects/weapons (but aircraft don't have death weapons so its less noticeable). They shouldn't do this because it looks bad.

There is another bug that I'm unsure of the best way to fix:
- transports/carriers can't be ctrl-k'd. 
This bug is caused because `Entity:Kill()` is separate from `Entity:Kill(instigator, damageType, excessDamageRatio)`, and the latter is called if any 3 arguments are passed to :Kill, even if they are nil, and the arguments must be valid (`instigator is Unit | nil, damageType is string, excessDamageRatio is number`).
Solutions: Add `(nil, '', 0)` to selfdestruct.lua:45, add `if not instigator then damageType = '' excessDamageRatio = 0 end` to the AirTransportUnit and AircraftCarrierUnit `Kill` functions, or use BaseUnitKill(self) when instigator == nil inside the unit classes and inside killcargo. 


The best fix would just be whatever is most easily documented I guess.

Future? possible improvements:
- wrecks from transports' internal storages
- can an internal storage + external clamps transport exist? If so, would it break the logic?

Don't forget the changelog by the way.

clyfordv commented on lua/sim/units/components/TransportUnitComponent.lua:
```diff
@@ -59,30 +65,78 @@ BaseTransport = ClassSimple {
+        for _, unit in self.cargoCache or {} do
+            if not unit:BeenDestroyed() then
+                unit.DeathWeaponEnabled = false -- Units at this point have no weapons for some reason. Trying to fire one crashes the game.
+                unit:OnKilled(self.killInstigator, self.killDamageType, self.killExcessDamageRatio)
```
Those values aren't passed to/aren't available to OnImpact, so cannot be passed ImpactCargo, and need to be cached from Kill/OnKilled.

clyfordv replied:
As to overkill, would it not also be odd to see the wreck of the transport disappear but not the units?

lL1l1 replied:
By "passed in before" I meant why didn't the old devs cache it. I think it's just because that was the easiest solution at the time, because I can't find any old transport logic that would explain not caching as a workaround for something. Consider that question resolved.

For overkill, units have way more HP than transports, so aesthetically they can survive the complete destruction of the transport. After that, what kills them is the crash damage, which doesn't seem like it would obliterate the unit. In addition to all that, any low HP units that could be overkilled by AA would have already died while the transport was being shot at.

clyfordv replied:
Ah, I see I see. Yeah who knows, sometimes an obvious solution conceals itself.

And I can get with that reasoning. How would you feel about the overkill-surviving units being thrown away from the explosion?

lL1l1 replied:
I think it's best for the reclaim to fly in the same direction as the transport, so that the actions of the players is reflected in where the reclaim goes. The reclaim being next to the transport also helps remember what was in the transport for replay watchers/casters/in-game players.

clyfordv commented on lua/sim/units/components/TransportUnitComponent.lua:
```diff
@@ -59,30 +65,78 @@ BaseTransport = ClassSimple {
 
         local cargo = self:GetCargo()
+
+        -- If we have storage slots, don't cache our cargo, because it's stored inside and
+        -- we'll never have to deal with it again
+        local cacheCargo = (self:GetBlueprint().Transport.StorageSlots and false) and true
```
Sounds good to me, hadn't looked at the blueprint viewer.

clyfordv commented on lua/sim/units/components/TransportUnitComponent.lua:
```diff
@@ -59,30 +65,78 @@ BaseTransport = ClassSimple {
+            -- the engine will allegedly handle actually killing the unit, but misses some, so we'll
+            -- explicitly kill our unit (with an instigator) to avoid units slipping through the cracks
+            -- (and so the engine can properly update kill counts, both for score and unit kills)
+            unit:Kill(instigator, damageType, excessDamageRatio)
```
Your fix above addresses this ye? (And then uncached cargo gets killed, but never fires OnKilled. Will that prevent anything important?)

lL1l1 replied:
Not firing OnKilled is a problem, it does tons of things from visual effects to callbacks to statistics to AI informing.
Bug: It also disperses veterancy, so if the unit was ever damaged before, it gives vet to the transport killer (through the transport component) and to whatever damaged it before (double vet). This is a common situation since most ground AA has AoE.
I think the better way to handle dispersing unit veterancy of transported units is to concatenate the transport's veterancy fields (`VeterancyComponent.VetDamage, .VetInstigators, and .VetDamageTaken` onto the cargo's vet fields. Fixes double vet dispersal, but still gives double vet damage for splash AA, it could be accounted for by not adding the damage if that entity already has damage on the unit.
OnKilled needs a way to not trigger death visual effects. Maybe by using a 'TransportDeath' damage type (passes to AI brain), just adding a new optional parameter to the function, or by using a unit field similar to `Unit.DeathWeaponEnabled`. Btw air units have their own OnKilled behavior that happens when they're in the air layer (inside the carrier, even underwater atlantis, they are still in the air layer).

clyfordv replied:
I like the 'TransportDeath' damage type, it's a tight solution.

For the double veterancy issue, we can replace:

`cargoMass = cargoMass + unit:GetTotalMassCost()`
with
`cargoMass = cargoMass + unit:GetTotalMassCost() * (unit:GetHealth() / unit:GetMaxHealth())`

... which will mean that only any remaining health will be credited to the transport instigator, but it will still be divided properly among the transport killers.

clyfordv replied:
Working through an example:
1. `Tank` has 100 hp.
2. It gets in a scuffle, takes 25 hp of damage from `kbot`.
3. It loads in `transport` that has 50 hp.
4. Transport gets whacked by `flak`, `tank` takes 25 damage, `transport` takes 25 damage.
5. `Interceptor` shoots down transport.

Transport veterancy dispersal:
- 50% to `flak`
- 50% to `interceptor`

Tank veterancy dispersal:
- 25% to `kbot`
- 25% to `flak`
- 50% remaining when killed on transport

The leftover 50% of `tank:MassCost()` gets collected by `cargoMass` and dispersed 50:50 to `flak` and `interceptor` in `KillCargo()`

clyfordv commented on lua/sim/units/components/TransportUnitComponent.lua:
```diff
@@ -59,30 +65,78 @@ BaseTransport = ClassSimple {
+            if EntityCategoryContains(categories.TRANSPORTATION, unit) then
+                cargoMass = cargoMass + unit:KillCargo(instigator,  damageType, excessDamageRatio, false)
             end
-            unit:DetachFrom()
```
I'd asked the same question, couldn't think of a good reason why it had to be in there, and figured I'd experiment with taking it out.

lL1l1 replied:
It works fine and I can't find anything through git blame, so I guess it's okay.

clyfordv commented:
Fixed ctrl-k, probably had something to do with removing the part with comments to the effect of "do not remove this part or it won't work".

MrRowey commented:
Please Request a review from me once PR is Complete :)

clyfordv commented:
Alright we should be solid (tested it this time!) 

Improvements:
- Added `TransportInternal` and `TransportExternal` as damage types (currently only `TransportInternal` is relevant, so we could trim it down to just `Transport` if people like the simplicity).
- Reworked the logic so KillCargo only runs once on transports-in-transports. 
- OnKilled for mobile units and aircraft runs a check for the `TransportInternal` damage type and will not play any effects, death threads, shield bounce projectiles, etc. when the internal storage they're in dies.

Possible issues:
- The divide between score/AI and visual/audio/etc. in `OnKilled` is not super strong, or at least not formalized. (For example, do death threads do anything that should still happen if killed inside a transport?)

Minor issues:
- Discovered that external factory dummy units are counted as losses to the player at the end of the game (2, actually--1 structure, 1 layer). This happens in `deploy/fafdevelop` as well.

Future ideas:
- Tangentially related, but when a transport enters a transport it could detach its unit, store it in the carrier on another bone, then retrieve it when it exits. This would require a little rework of the above, but would solve the ghost selection issue when a stinger with cargo is in a carrier.

MrRowey commented:
@relent0r will any of the changes here affect ai ?

lL1l1 reviewed (changes requested):
I tested it and found no errors.

clyfordv commented on lua/sim/Unit.lua:
```diff
@@ -1541,6 +1545,11 @@ Unit = ClassUnit(moho.unit_methods, IntelComponent, VeterancyComponent) {
 
         self.Brain:OnUnitKilled(self, instigator, type, overkillRatio)
+
+        -- If we're in internal storage, we're done, destroy the unit to avoid any errors
+        if type == "TransportDamage" then
+            self:Destroy()
+        end
```
"TransportDamage" is for internal storage (external storage just gets "Normal"), so no death threads, wreckage, etc. "Any errors" == any lingering references (like from `OnImpact`, though that's applicable for aircraft only) that might throw warnings. Adding the `self:Destroy()` here (to mobile units) is just a precaution for something unforseen, because I haven't found anything that would be an issue but felt it wise to get a rid of a unit that would otherwise be hanging around with half its normal death declarations.

clyfordv commented on lua/sim/units/MobileUnit.lua:
```diff
@@ -95,9 +95,10 @@ MobileUnit = ClassUnit(Unit, TreadComponent) {
-            self.killedInTransport = false
+        -- Skips a single OnKilled call
+        -- currently used by transports with external storage, so that death effects can be applied later from OnImpact
+        if self.SkipOnKilled then
+            self.SkipOnKilled = false
```
I'll flip it back.

lL1l1 replied:
I think it should be changed back because otherwise any old mod that creates a similar OnKilled function will have to 
be unnecessarily updated to handle the new transport flag name.

clyfordv commented on lua/sim/units/AirTransportUnit.lua:
```diff
@@ -116,10 +116,18 @@ AirTransport = ClassUnit(AirUnit, BaseTransport) {
+            self:KillCargo(instigator)
+        end
+        -- these need to be defined for certain behaviors (like ctrl-k) to function
+        damageType = damageType or "Normal"
```
If it works it works, I was just going with what was defined in the original.

clyfordv commented on lua/sim/units/components/TransportUnitComponent.lua:
```diff
@@ -66,63 +69,72 @@ BaseTransport = ClassSimple {
+        if recursive then
             return cargoMass
+        else
+            self:VeterancyDispersal(cargoMass)
```
It would not be difficult to do, but I think 1 level at a time is a good rate limiter. The extreme case of distributing it in chunks results in "kill a fully loaded carrier" -> "immediately promoted to top rank", which feels odder to me, given that the actual action of shooting down the transport is the same as the empty one.

E.g. In real life(!), a dive bomber pilot that gets a lucky hit on a loaded carrier or AA crew that shoots down an airship would probably get a bonus/promotion/three day pass, but would likely not be promoted to admiral/general.

clyfordv commented on lua/armordefinition.lua:
```diff
@@ -29,6 +29,7 @@
 ---| "TreeFire"
 ---| "TreeForce"
 ---| "WallOverspill"
+---| "TransportDamage"
```
I think we're alright (conceptually there's a good number of damage types that are used for various kinds of bookkeeping, not just damage) but it also prevents us from having to declare 150 new table fields for a single instant when a fully loaded CZAR gets killed.

relent0r commented:
> @relent0r will any of the changes here affect ai ?

I don't see anything that would impact the default AI. The changing from OnKilled to Kill might impact other AI's that are tracking deaths of units. But it looks like we are trying to make it consistent between files so makes sense.

MrRowey commented:
@lL1l1 you happy with the changes so far? if so can you approve the PR so I know

clyfordv commented:
killedInTransport flagging variable is renamed back, so we should be squared away
